Add support for stream resources in FinfoMimeTypeDetector::detectMimeType()

das-peter opened the pull request:
As far as I can tell it is valid to call that method with a resource. As such I think it's worth to attempt to detect the mime-type by the stream contents too.
This should be viable via stream_get_contents() while maintaining the stream pointers using ftell() and fseek().

frankdejonge commented on src/FinfoMimeTypeDetector.php:
```diff
@@ -53,9 +53,14 @@ public function __construct(
+            $mimeType = @$this->finfo->buffer($this->takeSample($contents));
+        } elseif (is_resource($contents) && get_resource_type($contents) === 'stream') {
+            $streamPosition = ftell($contents);
+            $mimeType = @$this->finfo->buffer(stream_get_contents($contents, $this->bufferSampleSize, 0));
```
The bufferSampleSize can be null, but should have a default in this method call. I'm not sure what it should be. In the docs of node's file-type they show a value of 4100 which seems a bit much but the docs are written by an industry veteran, so might be ok as the default since it would be the entire file if not set.

das-peter replied:
Adjusted - and now using a temporary stream for the detection since stream_copy_to_stream(), contrary to stream_get_contents()), should not reserve the whole buffer if unnecessary.
I'm currently torn between using `php://memory` and `php://temp` - the temp stream would have larger buffer support by eventually switching from memory to file handling.
However, it's cut-off is 2MB - which is larger than the now defined default.

das-peter replied:
Switched to `php://temp` with custom memory cut-off limit set to the default buffer length. This way large custom buffers should not lead to memory hogging while the default will benefit from no file system operations.

frankdejonge commented on src/FinfoMimeTypeDetector.php:
```diff
@@ -53,9 +53,14 @@ public function __construct(
+        $mimeType = null;
+        if (is_string($contents)) {
+            $mimeType = @$this->finfo->buffer($this->takeSample($contents));
+        } elseif (is_resource($contents) && get_resource_type($contents) === 'stream') {
```
The `if` condition should include a check to see if the stream is seekable, otherwise the calls below result in an error.

das-peter replied:
Addressed by generally ignoring non-seekable streams, while providing an option to change this behavior when creating the class.

das-peter commented:
@frankdejonge Thank you very much for your prompt response!
I've tried to address your valuable feedback.
The updated version has the stream sample extraction isolated in a dedicated method as it got complex.

frankdejonge commented on src/FinfoMimeTypeDetector.php:
```diff
@@ -90,6 +105,48 @@ private function takeSample(string $contents): string
+                return '';
+            }
+
+            // Memory optimization: given a length stream_get_contents()
```
where can I validate this?

das-peter replied:
https://www.php.net/manual/en/function.stream-get-contents.php#refsect1-function.stream-get-contents-notes

frankdejonge commented on src/FinfoMimeTypeDetector.php:
```diff
@@ -39,23 +39,38 @@ class FinfoMimeTypeDetector implements MimeTypeDetector, ExtensionLookup
+        $mimeType = null;
+        if (is_string($contents)) {
+            $mimeType = @$this->finfo->buffer($this->takeSample($contents));
+        } elseif (is_resource($contents)) {
```
Instead of detecting if the stream is seekable in `takeSample` let's extract it and us it in the if condition. Basically, I'd want the sampling only to happen when the stream is seekable, otherwise get the current behaviour, otherwise it's a BC break and I don't think this warrants it.

das-peter replied:
Done. However, not sure if the notation should be shortened to `!empty(stream_get_meta_data($contents)['seekable'])` - am not sure about php behavior in all supported versions nor do I know your personal preference :).

frankdejonge commented on src/FinfoMimeTypeDetector.php:
```diff
@@ -39,23 +39,38 @@ class FinfoMimeTypeDetector implements MimeTypeDetector, ExtensionLookup
         $this->extensionMap = $extensionMap ?: new GeneratedExtensionToMimeTypeMap();
         $this->bufferSampleSize = $bufferSampleSize;
         $this->inconclusiveMimetypes = $inconclusiveMimetypes;
+        $this->ignoreNonSeekableStreams = $ignoreNonSeekableStreams;
```
with the other suggestion implemented, this configuration option should be removed

das-peter replied:
Done. Tests adjusted accordingly.
Currently unhappy about using `http://` as a non-seekable stream but couldn't come up with a better idea as of yet.
Do you by chance know if there's some phpunit stream mocks that could do that?

das-peter commented:
@frankdejonge Thanks again for your responsiveness. I think I've addressed your points, I'll let you "resolve" the conversations as you'll be the judge if the solutions are satisfactory.

das-peter commented:
Allowed myself to resolve the reviews in the hope to progress this further
